docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+It is *not* a goal to import every C++ API into Swift. APIs that are imported by Swift should feel Swifty and idiomatic. There must be an easy way to add annotations so that APIs become importable, especially when updating APIs directly is not an option (because the source code is not available). 
+
+## The approach
+
+The approach for importing C++ APIs into Swift is derived from the fact that both our goals, and C++ as a language, necessitate that not all APIs are imported, and to import an API the user must provide semantic information in some cases. Further, in order to realize our goals (especially the goals of safety and clear, well defined mappings), it is impossible to import any arbitrary C++ API. APIs that are imported by Swift must fall into one of our understood categories or be annotated with semantic information that can be used to classify them into a Swift idiom. 
+
+In this section, the roadmap will discuss several of the most common C++ API patterns and how they might be imported by the Swift compiler using the goals outlined above.
+
+### Definitions
+
+**Reference types** have reference semantics and object identity. A reference type is a pointer (or “reference”) to some object which means there is a layer of indirection. When a reference type is copied, the pointer’s value is copied rather than the object’s storage. This means reference types can be used to represent non-copyable types in C++.
+
+**Owned types** “own” some storage which can be copied and destroyed. An owned type must be copyable and destructible. The copy constructor must copy any storage that is owned by the type and the destructor must destroy that storage. Copies and destroys must balance out and these operations must not have side effects. Examples of owned types include `std::vector` and `std::string`. 
+
+**An iterator** represents a point in a range. Iterator types must provide a comparison operator and increment operator. Iterators can be returned by `begin` and `end` methods to form a range.
+
+**Trivial types** can be copied by copying the bits of a value of the trivial type and do not need any special destruction logic. Trivial types own their storage.
+
+**Projections** are references to storage. Projections of owned types are usually unsafe. Most iterators are projections. 
+
+
+### Reference types
+
+Swift types fall into two categories: value types and reference types. The same is true for Objective-C types, which makes importing types easy: structs are imported as value types and Objective-C classes are imported as Swift class types. The same is not true for C++. In typical C++ fashion, there is no clear idiom for defining reference types. All types in C++ are declared in the same way, but some types have reference semantics and others don’t. To be able to express reference types in Swift natively, users must annotate their reference types as such, which will tell the compiler to import them as Swift classes (which have the same semantics). 
```

Comment:
This is not really a distinction between Swift and C++.  I can create `struct`s with reference semantics and `class`es with value semantics in Swift.

Reply:
Also, Swift classes do not have the same semantics as C++ reference types.  Typically when C++ is used with reference semantics, types are handled through explicit pointers, so `shared_ptr<Foo>` (or `Foo*`) is treated as a reference to `Foo` even if `Foo` itself would otherwise have value semantics.  Idiomatic C++ actually gives you quite a bit more information than idiomatic Swift does in this respect.

Reply:
Pls explain how it's possible to create `struct` with reference semantics in Swift, as from my understanding, Swift's `struct` always "behaves" as if it was just a value type, i.e. it's copied when passed/dereferenced. Under the hood there might be some optimizations, such as copy-on-write in collections, for example, but at the surface level structs always behave like value types.

... unless there are `mutating` functions and/or `inout` parameters, but those apply to functions and not to the structs themselves. Or am I missing something else?